docs/proposals/job.md

```diff
@@ -0,0 +1,145 @@
+# Job Controller
+
+## Abstract
+A proposal for implementing a new controller - Job controller - which will be responsible
+for managing pod(s) that require to run-once to a completion, in contrast to what
+ReplicationController currently offers.
+
+Several existing issues and PRs were already created regarding that particular subject:
+* Job Controller [#1624](https://github.com/GoogleCloudPlatform/kubernetes/issues/1624)
+* New Job resource [#7380](https://github.com/GoogleCloudPlatform/kubernetes/pull/7380)
+
+
+## Use Cases
+1. Be able to start of one or several pods tracked as a single entity.
+1. Be able to implement basic batch oriented tasks.
+1. Be able to get the job status.
+1. Be able to limit the execution time for a job.
+1. Be able to specify the number of instances performing a job.
+1. Be able to specify triggers on job’s success/failure.
+
+
+## Motivation
+Jobs are needed for executing multi-pod computation to completion; a good example
+here would be the ability to implement any type of batch oriented tasks or a MapReduce
+or Hadoop style workload.
+
+
+## Implementation
+Job controller is similar to replication controller in that they manage pods.
+This implies they will follow the same controller framework that replication
+controllers already defined.  The biggest difference between a `Job` and a
+`ReplicationController` object is the purpose; `ReplicationController`
+ensures that a specified number of Pods are running at any one time, whereas
+`Job` is responsible for keeping the desired number of Pods to a completion of
+a task. This will be represented by the `RestartPolicy` which is required to
+always take value of `RestartPolicyNever` or `RestartOnFailure`.
+
+
+The new `Job` object will have the following content:
+
+```go
+// Job represents the configuration of a single job.
+type Job struct {
+    TypeMeta
+    ObjectMeta
+
+    // Spec is a structure defining the expected behavior of a job.
+    Spec JobSpec
+
+    // Status is a structure describing current status of a job.
+    Status JobStatus
+}
+
+// JobList is a collection of jobs.
+type JobList struct {
+    TypeMeta
+    ListMeta
+
+    Items []Job
+}
+```
+
+`JobSpec` structure is defined to contain all the information how the actual job execution
+will look like.
+
+```go
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
```

Comment:
@davidopp I think that's the wrong point of control. If I understand you, I would be dynamically adding to the number of completions as the job continued to run. It would be extremely racy. What if all tasks transitioned to phase complete (and thus the job transitioned to phase complete) before my controller had a chance to increment add to the number of completions. Can a phase complete job transition to a phase running job? If I wanted a job to run with MaxParallelism=1, this scenario would happen every time. This is not impossible to implement but It seems really nasty.

If I were to design a controller that does what the example describes, it would operate on Task not Job.